src/System.CommandLine/Help/HelpBuilder.cs

```diff
@@ -204,103 +206,199 @@ private void AppendDescription(string description)
+                throw new ArgumentNullException(nameof(helpItems));
 
-            var offset = maxInvocationWidth + ColumnGutter - helpItem.Invocation.Length;
-            var availableWidth = GetAvailableWidth();
-            var maxDescriptionWidth = availableWidth - maxInvocationWidth - ColumnGutter;
-            var descriptionColumn = helpItem.Description;
-            if (helpItem.HasDefaultValueHint)
+            var table = CreateTable(helpItems, item => new[]
             {
-                var postfix = string.IsNullOrEmpty(helpItem.Description) ? string.Empty : " ";
-                descriptionColumn += postfix + helpItem.DefaultValueHint;
-            }
-            var descriptionLines = SplitText(descriptionColumn, maxDescriptionWidth);
-            var lineCount = descriptionLines.Count;
+                item.Invocation,
+                JoinNonEmpty(" ", item.Description, item.DefaultValueHint)
+            });
 
-            AppendLine(descriptionLines.FirstOrDefault(), offset);
+            var columnWidths = ColumnWidths(table);
+            AppendTable(table, columnWidths);
+        }
 
-            if (lineCount == 1)
-            {
-                return;
+        /// <summary>
+        /// Create a read only table of strings using the projection of a collection.
+        /// </summary>
+        /// <typeparam name="T">The type of the elements of <paramref name="collection"/>.</typeparam>
+        /// <param name="collection">The collection of values to create the table from.</param>
+        /// <param name="selector">A transformation function to apply to each element of <paramref name="collection"/>.</param>
+        /// <returns>
+        /// A read only table of strings whose elements are the projection of the collection with whitespace formatting removed.
+        /// </returns>
+        protected virtual ReadOnlyCollection<ReadOnlyCollection<string>> CreateTable<T>(IEnumerable<T> collection, Func<T, IEnumerable<string>> selector)
+        {
+            return collection.Select(selector)
+                .Select(row => row
+                    .Select(element => RemoveFormatting(element))
+                    .ToList().AsReadOnly())
+                .ToList().AsReadOnly();
+        }
+
+        /// <summary>
+        /// Allocate space for columns favoring minimal rows. Wider columns are allocated more space if it is available.
+        /// </summary>
+        /// <param name="table">The table of values to determine column widths for.</param>
+        /// <returns>A collection of column widths.</returns>
+        protected virtual ReadOnlyCollection<int> ColumnWidths(IEnumerable<IList<string>> table)
+        {
+            if (!table.Any())
+                return Array.AsReadOnly(Array.Empty<int>());
+
+            var columns = table.First().Count;
+            Debug.Assert(table.All(e => e.Count == columns), $"Every row in {nameof(table)} must have the same number of columns.");
```

Comment:
Please remove the `Debug.Assert`.

Reply:
I can remove the `Debug.Assert` but if `ColumnWidths` is called with an unequal number of columns an exception will be thrown at an indeterminate time. This should never happen in production so I thought an O(n) check for this was generally wasteful.

Reply:
Would a compensation of some kind be preferable to an exception? Or can we make the incorrect state unrepresentable by the object model?

Reply:
If the object model was a multidimensional array incorrect state would be unrepresentable as it would guarantee a consistent number of columns.

`CreateTable` will always return a table with a consistent number of columns but if `ColumnWidths` is called with arbitrary input then this cannot be guaranteed.